### PR TITLE
F: BatchToIntacct - Add support for debit specific attributes

### PR DESCRIPTION
**Pull request by guymaness2**
##### What does the change add or fix?

Added new debit account specific classification attributes and support for grouping by them. Also added support for grouping by custom dimensions when using a GL account grouping mode.

**New Attributes:**

- Default Debit Project
- Debit Class
- Debit Department
- Debit Location

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

See https://github.com/KingdomFirst/RockBlocks/pull/147

---------

##### Who reported, requested, or paid for the change?

The Father's House

---------

##### Does this update or add options to the block UI?

Financial Account Attribute Changes:  
image
  
---------

##### What files does it affect?

* rocks.kfs.Intacct/IntacctJournal.cs
* rocks.kfs.Intacct/IntacctModels.cs
* rocks.kfs.Intacct/IntacctOtherReceipt.cs
* rocks.kfs.Intacct/Migrations/001_AddSystemData.cs
* rocks.kfs.Intacct/Migrations/002_AddTransactionFeeAttributes.cs
* rocks.kfs.Intacct/Migrations/003_AddBatchesToIntacct.cs
* rocks.kfs.Intacct/Migrations/004_AddDebitAccountAttributes.cs
* rocks.kfs.Intacct/Properties/AssemblyInfo.cs
* rocks.kfs.Intacct/SystemGuid/Attribute.cs
* rocks.kfs.Intacct/Utils/TransactionHelpers.cs
* rocks.kfs.Intacct/rocks.kfs.Intacct.csproj

---------

##### Is it a breaking change for other versions/clients?

no

### CODE REVIEW

File: rocks.kfs.Intacct/Utils/TransactionHelpers.cs [hunk 1]
```diff
@@ -207,6 +228,28 @@ public static List<string> GetCustomDimensions()
 
             return mergeFields;
         }
+
+        public static SortedDictionary<string, dynamic> GetFilteredDimensions( SortedDictionary<string, dynamic> customDimenesionValues, string suffixExclude = null, string suffixRemove = null )
+        {
+            var filteredDimensions = new SortedDictionary<string, dynamic>();
+            if ( suffixExclude.IsNotNullOrWhiteSpace() )
+            {
+                customDimenesionValues = new SortedDictionary<string, dynamic>( customDimenesionValues.Where( d => !d.Key.ToLower().EndsWith( suffixExclude ) ).ToDictionary( x => x.Key, x => x.Value ) );
```
Comment (guymaness2) [inline rocks.kfs.Intacct/Utils/TransactionHelpers.cs:237]:
```suggestion
                customDimenesionValues = new SortedDictionary<string, dynamic>( customDimenesionValues.Where( d => !d.Key.ToLower().EndsWith( suffixExclude.ToLower() ) ).ToDictionary( x => x.Key, x => x.Value ) );
```

File: rocks.kfs.Intacct/Utils/TransactionHelpers.cs [hunk 2]
```diff
@@ -207,6 +228,28 @@ public static List<string> GetCustomDimensions()
...
+            }
+
+            foreach ( var dimension in customDimenesionValues )
+            {
+                if ( suffixRemove.IsNotNullOrWhiteSpace() && dimension.Key.ToLower().EndsWith( suffixRemove.ToLower() ) )
+                {
+                    filteredDimensions.Add( dimension.Key.Remove( dimension.Key.Length - suffixRemove.Length ), dimension.Value );
```
Comment (guymaness2) [inline rocks.kfs.Intacct/Utils/TransactionHelpers.cs:244]: test

File: rocks.kfs.Intacct/IntacctJournal.cs [hunk 3]
```diff
@@ -267,41 +271,50 @@ private List<JournalEntryLine> GenerateLineItems( List<GLBatchTotals> transactio
...
                         Amount = s.Sum( f => f.Amount ),
                         DebitAccount = s.Key.DebitAccount,
-                        Class = s.Key.Class,
-                        Department = s.Key.Department,
-                        Location = s.Key.Location,
-                        Project = s.Key.Project,
+                        DebitClass = s.Key.DebitClass,
+                        DebitDepartment = s.Key.DebitDepartment,
+                        DebitLocation = s.Key.DebitLocation,
+                        CreditProject = s.Key.DebitProject,
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctJournal.cs:302]:
```suggestion
                        DebitProject = s.Key.DebitProject,
```

File: rocks.kfs.Intacct/IntacctJournal.cs [hunk 4]
```diff
@@ -267,41 +271,50 @@ private List<JournalEntryLine> GenerateLineItems( List<GLBatchTotals> transactio
...
                         Description = s.First().Description,
                         CustomDimensions = s.First().CustomDimensions
                     } )
                     .ToList();
 
                 feeDebitTransactions = feeDebitTransactions
-                    .GroupBy( d => new { d.Class, d.Department, d.Location, d.Project, d.DebitAccount } )
+                    .GroupBy( d => new { d.DebitClass, d.DebitDepartment, d.DebitLocation, d.DebitProject, d.DebitAccount } )
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctJournal.cs:309]:
```suggestion
                    .GroupBy( d => new { d.DebitClass, d.DebitDepartment, d.DebitLocation, d.DebitProject, d.CustomDimensionString, d.DebitAccount } )
```

File: rocks.kfs.Intacct/IntacctJournal.cs [hunk 5]
```diff
@@ -312,41 +325,49 @@ private List<JournalEntryLine> GenerateLineItems( List<GLBatchTotals> transactio
...
+                        CreditLocation = s.Key.CreditLocation,
+                        CreditProject = s.Key.CreditProject,
                         Description = s.First().Description,
                         CustomDimensions = s.First().CustomDimensions
                     } )
                     .ToList();
 
                 feeCreditTransactions = feeCreditTransactions
-                    .GroupBy( d => new { d.Class, d.Department, d.Location, d.Project, d.CreditAccount } )
+                    .GroupBy( d => new { d.CreditClass, d.CreditDepartment, d.CreditLocation, d.CreditProject, d.CreditAccount } )
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctJournal.cs:362]:
```suggestion
                    .GroupBy( d => new { d.CreditClass, d.CreditDepartment, d.CreditLocation, d.CreditProject, d.CreditAccount, d.CustomDimensionString } )
```

File: rocks.kfs.Intacct/IntacctOtherReceipt.cs [hunk 6]
```diff
@@ -264,7 +264,7 @@ private OtherReceipt BuildOtherReceipt( FinancialBatch financialBatch, ref strin
                     Memo = DescriptionLava.ResolveMergeFields( mergeFields ),
                     LocationId = locationId,
                     DepartmentId = departmentId,
-                    ProjectId = bTran.Project,
+                    ProjectId = bTran.CreditProject,
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctOtherReceipt.cs:267]: Make sure custom dimensions work properly in OtherReceipts

File: rocks.kfs.Intacct/Migrations/004_AddDebitAccountAttributes.cs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+// limitations under the License.
+// </copyright>
+//
+using System;
+using Rock.Plugin;
+using KFSConst = rocks.kfs.Intacct.SystemGuid;
+
+namespace rocks.kfs.Intacct.Migrations
+{
+    [MigrationNumber( 4, "1.11.0" )]
```
Comment (guymaness2) [inline rocks.kfs.Intacct/Migrations/004_AddDebitAccountAttributes.cs:23]:
```suggestion
    [MigrationNumber( 4, "1.13.0" )]
```

File: rocks.kfs.Intacct/Migrations/004_AddDebitAccountAttributes.cs [hunk 8]
```diff
@@ -0,0 +1,264 @@
+﻿// <copyright>
+// Copyright 2023 by Kingdom First Solutions
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//9DF341B7-2E07-44BA-94B1-F637BE551961
```
Comment (guymaness2) [inline rocks.kfs.Intacct/Migrations/004_AddDebitAccountAttributes.cs:7]:
```suggestion
// 
```

File: rocks.kfs.Intacct/Utils/TransactionHelpers.cs [hunk 9]
```diff
@@ -59,22 +59,37 @@ public static List<GLTransaction> GetTransactionSummary( FinancialBatch financia
...
                         transactionFeeAccount = gatewayDefaultFeeAccount;
                     }
 
-                    var projectCode = string.Empty;
+                    var creditProjectCode = string.Empty;
+                    var debitProjectCode = string.Empty;
                     if ( detailProject != null )
                     {
-                        projectCode = DefinedValueCache.Get( ( Guid ) detailProject ).Value;
+                        creditProjectCode = DefinedValueCache.Get( ( Guid ) detailProject ).Value;
```
Comment (guymaness2) [inline rocks.kfs.Intacct/Utils/TransactionHelpers.cs:75]: Make so detail project sets both

File: rocks.kfs.Intacct/Utils/TransactionHelpers.cs [hunk 10]
```diff
@@ -189,7 +206,11 @@ public static List<string> GetCustomDimensions()
                 foreach ( var rockKey in mergeFieldObjects.CustomDimensions )
                 {
                     var dimension = rockKey.Split( '.' ).Last();
-                    customDimensionValues.Add( dimension, account.GetAttributeValue( rockKey ) );
+                    var value = account.GetAttributeValue( rockKey );
+                    if ( value.IsNotNullOrWhiteSpace() )
+                    {
+                        customDimensionValues.Add( dimension, account.GetAttributeValue( rockKey ) );
```
Comment (guymaness2) [inline rocks.kfs.Intacct/Utils/TransactionHelpers.cs:212]:
```suggestion
                        customDimensionValues.Add( dimension, value );
```

File: rocks.kfs.Intacct/IntacctOtherReceipt.cs [hunk 11]
```diff
@@ -171,13 +171,21 @@ public XmlDocument CreateOtherReceiptXML( IntacctAuth AuthCreds, int BatchId, re
...
+                            //    writer.WriteStartElement( "customfields" );
+                            //    foreach ( KeyValuePair<string, dynamic> customField in item.CustomFields )
+                            //    {
+                            //        writer.WriteStartElement( "customfield" );
+                            //        writer.WriteElementString( "customfieldname", customField.Key );
+                            //        writer.WriteElementString( "customfieldvalue", customField.Value ?? string.Empty );
+                            //        writer.WriteEndElement();  // close customfield
+                            //    }
+                            //    writer.WriteEndElement();  // close customfields
+                            //}
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctOtherReceipt.cs:188]:
```suggestion

                            // Intacct api documentation shows support for custom fields, but we are unable to get them to work. Disabling for now.

                            //if ( item.CustomFields.Count > 0 )
                            //{
                            //    foreach ( KeyValuePair<string, dynamic> customField in item.CustomFields )
                            //    {
                            //        writer.WriteElementString( customField.Key, customField.Value ?? string.Empty );
                            //    }
                            //}
```

**Review by nateh777 — APPROVED** [review]

**Pull request merged**